### PR TITLE
TIEGCM 2.0 Manhattan

### LINKED ISSUE #309: TIEGCM
Notes for TIEGCM model_mod

X vs X_MN variables - Forcing X at the moment 25de28e77a731a5016ff57427cb53cdde841b0c8

Original note in Lanai model_mod.f90
```
! Note from FindVar_by_kind:
! FIXME There is some confusion about using the T-minus-1 variables
! in this construct. Both TN and TN_NM have the same dart_kind,
! so we use the first one ... but it is not guaranteed that TN
````

### PR DESCRIPTION
**Pull request by hkershaw-brown**
## Description:

TIEGCM model_mod updated to Manhattan
Tested with TIEGCM 2.0

* Supports 2.5-degree and 5-degree resolution grids.
* vtec extrapolation is done in model_mod::model_interpolate with 10 levels for 5deg, 20 levels for 2.5deg.
* Model_interpolate uses X rather than X_MN variable no matter the order of state
  variables in model_nml. Previously in Lanai the variable given first in the state was used.
  _MN variables are from the previous model time step.
* f10.7 can be added to the DART state. An example f10_7.cdl file is provided in the work directory. At the moment f10.7 is 
  in the CALCULATE/CONSTRUCTED domain. This could do with a better name. 
* Added hyperslice subroutine to state_structure_mod to read only part of a netcdf variable.
* Separated functions for io_dim_length and dim_length in state_structure_mod to
  allow for hyperslice.

Tiegcm has:
  - the lower boundary condition at the top level. This should not be altered by filter.
  - multiple timesteps per restart file.

**obs_def_upper_atm_mod.f90 changes**
  `get_expected_vtec` 'routine needs to be written' removed and replaced with common code.  

**state_structure_mod.f90 changes:**
new `hyperslice_domain` - cuts off the top level
new `has_unlimited_dim` query if the domain has an unlimited dimension.

`get_io_dim_length` -   returns io_info%dimlens so you can query the dimensions of the netcdf file vs. the state. 
`state_structure_info` print statement changed to not write out get_original_dim_ID - this segfaults for add_domain_from_spec.

**direct_netcdf_mod.f90 changes**
Read the latest slice of the unlimited dimension in the restart file.
Write the latest slice of the unlimited dimension in the restart file. 

**scripts**
There is are scripts in the directory in tiegcm/shell_scripts for running on Cheyenne
I am not sure if these are general enough to include in the release. I can remove them from this pull request. 

### Fixes issue

Fixes #309, #269

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

### Documentation changes needed?

- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.

### Tests
Please describe any tests you ran to verify your changes.

tiegcm model_mod_check
Cycled tiegcm on Cheyenne.  The shell scripts are included in this pull request,  I am not sure if these are general enough.  I can remove them from this pull request.

To build TIEGCM2.0 on Cheyenne I had to use esmf-7.1.0r. Cheyenne build script here: https://github.com/hkershaw-brown/tiegcm-cheyenne/blob/main/tiegcm-ch.job 

Tests for Netcdf changes:
lorenz_96 bitwise
wrf bitwise

## Checklist for merging

- [ ] Updated changelog entry
- [x] Documentation updated
- [ ] Version tag 

## Testing Datasets

- [x] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

### CODE REVIEW

File: assimilation_code/modules/io/direct_netcdf_mod.f90 [hunk 1]
```diff
@@ -852,22 +851,41 @@ subroutine read_variables(ncfile_in, var_block, start_var, end_var, domain)
...
+   num_dims = get_io_num_dims(domain, i)
+   allocate(counts(num_dims))
+   allocate(slice_start(num_dims))
 
-   dims = get_io_dim_lengths(domain, i)
+   if (has_unlimited_dim(domain)) then
+
+      slice_start(:) = 1 ! default to read all dimensions start at 1
+      counts(num_dims) = 1 ! one slice of unlimited dimesion
+      counts(1:num_dims-1) = get_dim_lengths(domain, i) ! the state
```
Comment (nancycollins) [inline assimilation_code/modules/io/direct_netcdf_mod.f90:863]: a couple of comments for the future here.  newer versions of netcdf allow for multiple unlimited dimensions so this approach will have to be altered if we run into a model that uses this in the input/output state to dart.  we can always require a model_to_dart and dart_to_model as a temporary measure once we do see how people might be using multiple unlimited dims.

also, it's possible that netcdf variables in files have additional (not-unlimited) dimensions that shouldn't be part of the state.  i can't remember which example or test had this, so for now it's not an issue.   but eventually some namelist items may be needed to address i/o issues - for example, the name(s) of the unlimited dimension(s) (e.g. Time), other name(s) of dimensions to ignore when reading into the state (e.g. leapfrog_step1 or something), and then for the ignored dimensions, what index should be used if the length is more than 1 (e.g. last, 1, etc).

File: assimilation_code/modules/io/state_structure_mod.f90 [hunk 2]
```diff
@@ -936,7 +947,76 @@ subroutine load_common_cf_conventions(domain)
...
+
+domain_size = 0
+! variable size
+do ivar = 1, get_num_variables(dom_id)
+
+   variable_size = 1
+
+   do jdim = 1, get_num_dims(dom_id, ivar)
+      !HK there is a todo about this, what is going on with time and member?
+      if ((state%domain(dom_id)%variable(ivar)%dimname(jdim) == 'time') .or. &
```
Comment (nancycollins) [inline assimilation_code/modules/io/state_structure_mod.f90:997]: i remember this being added to the existing code but i don't remember why, sorry.  "member" is the dimension created by dart when making a unified (single) file that holds all the ensemble member info.  "time" is most often an unlimited dim but nothing in netcdf forces it to be.

File: models/tiegcm/model_mod.f90 [hunk 3]
```diff
@@ -160,82 +161,52 @@ module model_mod
...
+real(r8)  :: bot_lat        = MISSING_R8
+real(r8)  :: top_lat        = MISSING_R8
+real(r8)  :: delta_lat      = MISSING_R8
+integer   :: zero_lon_index = MISSING_I
+
+
+! Obs locations are expected to be given in height [m],
+! and so vertical localization coordinate is *always* height.
+! Note that gravity adjusted geopotential height (ZG) is read in
+! "tiegcm_s.nc". ZG is 'cm', dart is mks(HK mks?))
```
Comment (nancycollins) [inline models/tiegcm/model_mod.f90:189]: "mks" = meter, kilogram, second as the standard units.

File: models/tiegcm/model_mod.f90 [hunk 4]
```diff
@@ -251,143 +222,36 @@ subroutine static_init_model()
...
+! error-check, convert namelist input to variable_table, and build the
+! state structure
+call verify_variables()
 
-! Might as well use the Gregorian Calendar
 call set_calendar_type('Gregorian')
 
-! Convert the last year/doy/hour/minute to a dart time.
-state_time = get_state_time(tiegcm_restart_file_name)
+! Convert the last year/day/hour/minute to a dart time.
```
Comment (nancycollins) [inline models/tiegcm/model_mod.f90:238]: does tiegcm time use 'doy' == day of year, 1-366?  if so, this isn't really a typo.  but i suppose day is just as clear, to be honest.

File: models/tiegcm/model_mod.f90 [hunk 5]
```diff
@@ -160,82 +161,52 @@ module model_mod
...
+real(r8)  :: bot_lon        = MISSING_R8
+real(r8)  :: top_lon        = MISSING_R8
+real(r8)  :: delta_lon      = MISSING_R8
+real(r8)  :: bot_lat        = MISSING_R8
+real(r8)  :: top_lat        = MISSING_R8
+real(r8)  :: delta_lat      = MISSING_R8
+integer   :: zero_lon_index = MISSING_I
+
+
+! Obs locations are expected to be given in height [m],
```
Comment (nancycollins) [inline models/tiegcm/model_mod.f90:186]: maybe adjust this comment?  it looks like model_interpolate also supports obs on model levels?  it's probably true that only height localization is supported, but asking for obs values on levels seems like it works?

File: models/tiegcm/model_mod.f90 [hunk 6]
```diff
@@ -396,371 +260,247 @@ function get_model_size()
...
-! Get lon and lat grid specs
-bot_lon        = lons(1)                         ! 180.
-delta_lon      = abs((lons(1)-lons(2)))          ! 5. or 2.5
-zero_lon_index = int(bot_lon/delta_lon) + 1      ! 37 or 73
-top_lon        = lons(nlon)                      ! 175. or 177.5
-bot_lat        = lats(1)                         !
-top_lat        = lats(nlat)                      !
-delta_lat      = abs((lats(1)-lats(2)))          !
+      ! HK pressure is not in the state so do not call get_state.
+      call error_handler(E_ERR, 'model_interpolate', 'vert is height for pressure interpolate')
```
Comment (nancycollins) [inline models/tiegcm/model_mod.f90:360]: if this is an error message, it needs to say why it's failing.  "pressure obs with vert in height not supported"  or something?

File: models/tiegcm/model_mod.f90 [hunk 7]
```diff
@@ -769,1225 +509,345 @@ end subroutine get_state_meta_data
...
+integer,  dimension(nmtime) :: mtime  ! day, hour, minute
+integer                     :: year, doy, utsec
+integer, allocatable, dimension(:,:) :: mtimetmp
+integer, allocatable, dimension(:)   :: yeartmp
 
-if ( .not. module_initialized ) call static_init_model
+if ( present(ncfileid) ) then
+   ncid = ncfileid
+else
+   call nc_check(nf90_open(filename, NF90_NOWRITE, ncid), &
```
Comment (nancycollins) [inline models/tiegcm/model_mod.f90:785]: eventually these calls could use the dart netcdf utility routines.

File: models/tiegcm/model_mod.f90 [hunk 8]
```diff
@@ -769,1225 +509,345 @@ end subroutine get_state_meta_data
...
-   call error_handler(E_MSG,'read_TIEGCM_namelist:',string1)
-endif
-
-end subroutine read_TIEGCM_namelist
-
-
-!-------------------------------------------------------------------------------
-
-
 subroutine read_TIEGCM_definition(file_name)
```
Comment (nancycollins) [inline models/tiegcm/model_mod.f90:851]: also eventually could be converted to using the dart netcdf util routines.

**Review by nancycollins — APPROVED** [review]
i left a few comments but nothing that prevents this from being merged with main.  i will suggest that the tiegcm test cases include obs locations in height that are exactly on model levels, both at the bottom, top, and in the middle to test when the fractions across the levels are exactly 0.0, 1.0, and inbetween.  but otherwise, getting rid of vector_to_model, model_to_vector, progvar, and moving up to the manhattan interfaces is a huge step forward.  

the release notes for this might ask users to report any changes in how dart reads their netcdf files.  i don't see how the state struct and i/o changes here would cause problems, but netcdf is so flexible and users are so creative in how they structure their data that it's hard to predict what changes might cause problems for someone.

File: assimilation_code/modules/io/direct_netcdf_mod.f90 (hunk 1 above)
Comment (hkershaw-brown) [reply assimilation_code/modules/io/direct_netcdf_mod.f90:863]: Totally agree with this, I'm keeping track State IO issues  here: https://github.com/NCAR/DART/issues/359 for a future refactor.

File: models/tiegcm/model_mod.f90 (hunk 4 above)
Comment (hkershaw-brown) [reply models/tiegcm/model_mod.f90:238]: oh, I read doy and day as the same word. 

tiegcm uses day,hour,minute. The tiegcm user guide uses 'day' for day-of-year, e.g.

```
STOP integer(3) Model stop time (day,hour,minute)
````

so it makes sense for DART to follow that.

File: models/tiegcm/model_mod.f90 (hunk 6 above)
Comment (hkershaw-brown) [reply models/tiegcm/model_mod.f90:360]: I'll strip this out, this is me catching pressure when testing.  Sorry.

File: models/tiegcm/model_mod.f90 (hunk 7 above)
Comment (hkershaw-brown) [reply models/tiegcm/model_mod.f90:785]: I'll do this in this pull request - good time to do it.

File: models/tiegcm/model_mod.f90 (hunk 8 above)
Comment (hkershaw-brown) [reply models/tiegcm/model_mod.f90:851]: ditto

File: models/tiegcm/model_mod.f90 (hunk 6 above)
Comment (hkershaw-brown) [reply models/tiegcm/model_mod.f90]: removed.

File: models/tiegcm/model_mod.f90 (hunk 5 above)
Comment (hkershaw-brown) [reply models/tiegcm/model_mod.f90]: done.

File: models/tiegcm/model_mod.f90 (hunk 7 above)
Comment (hkershaw-brown) [reply models/tiegcm/model_mod.f90]: done [cb492ec](https://github.com/NCAR/DART/commit/cb492ecf7a6469105de176a65a27660589180cde)

File: models/tiegcm/model_mod.f90 (hunk 8 above)
Comment (hkershaw-brown) [reply models/tiegcm/model_mod.f90:817]: done 8445726fee511fa19ecd2630a559a2dc2bdbd691

**Pull request merged**